### DISCUSSION TITLE
[Sionna RT] Contribution of precoder/combiner weights in Coverage Map function

### DISCUSSION THREAD

**Discussion by mauro-belgiovine** [NVlabs/sionna#150] [General]
Dear all,

I am exploring Sionna RT functionalities, I am especially interested in understanding how does the [`sionna.rt.Scene.coverage_map`](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Scene.coverage_map) works with respect to the `combining_vec` and `precoding_vec` parameters and how the [`sionna.rt.Scene.synthetic_array`](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Scene.synthetic_array) property interact with the coverage function.

Specifically, I would like to experiment with different precoders and combiners configurations applied at transmitter and receiver side and observe how the beam orientations (computed using a steering vector function) affects the channel power perceived at a specific location of the coverage map. My expectation is that, based on the beam configurations, the perceived power _should_ change at a given cell in the coverage map, although I did some tests with a 8x8 TX array and single antenna RX (both with `sionna.rt.Scene.synthetic_array` set to `True` or `False`) and it doesn't seem like there is a dramatic change in the perceived power when I change the precoder. Here is an example with the default precoding (i.e. 90 degree, perpendicular to the array) vs. a precoder computed using the steering vector function and pointing to a different angle (240 degrees) :
- precoder pointing at 90 degrees
![Unknown-2](https://github.com/NVlabs/sionna/assets/5329353/26a1974a-f3ed-4126-8804-c0e6af1fcf1a)
- precoder pointing at 240 degrees
![Unknown-3](https://github.com/NVlabs/sionna/assets/5329353/40affa03-9f81-4bee-85d8-9389bf375756)

I was wondering if I could find the answer to these questions in order to better debug this behavior:

1. Is there a function provided by Sionna to compute precoder/combiners, given an array configuration and an angle (i.e. steering function)?
2. What function should I use to visualize the radiation pattern of an array when a precoder/combiner is applied?

Thanks in advance for any advice, I have just started working with this library and I am excited to do some great simulations with it!

**Comment by faycalaa** [comment]
Hi,

There are currently no additional utilities for computing a steering vector or visualizing the radiation pattern when a precoder/combiner is applied.
Also, only synthetic arrays are used to compute the coverage map.

Here is a short piece of code that illustrates how a precoding vector changes the radiation pattern:

```py
scene = load_scene(sionna.rt.scene.munich)

num_cols = 64
scene.tx_array = PlanarArray(num_rows=1,
            num_cols=num_cols,
            vertical_spacing=0.5,
            horizontal_spacing=0.5,
            pattern="iso",
            polarization="V")

scene.rx_array = PlanarArray(num_rows=1,
                            num_cols=1,
                            vertical_spacing=0.5,
                            horizontal_spacing=0.5,
                            pattern="iso",
                            polarization="V")

# Transmitter
tx = Transmitter(name="tx",
                position=[35.11, 60.08, 10.],
                orientation=[0,0,0])
scene.add(tx)

# RX
rx = Receiver(name="rx",
              position=[95., 62., 10.0],
              orientation=[0,0,0])
scene.add(rx)
tx.look_at(rx)

# Steering vector
steering_angle = 20. # degrees
steering_angle = steering_angle*np.pi/180. # radian
precoding_vec = tf.range(num_cols, dtype=tf.float32)
precoding_vec = tf.exp(tf.complex(0., 2.*np.pi*precoding_vec*0.5*tf.math.sin(steering_angle)))
precoding_vec = precoding_vec / tf.linalg.norm(precoding_vec)

# Capture the power on a vertical coverage map
cm = scene.coverage_map(max_depth=0, cm_cell_size=[1.,1.], precoding_vec=precoding_vec, cm_size=[50., 100.], cm_orientation=[0.0, 0.5*np.pi, 0.0], cm_center=[95., 62., 0.])

scene.preview(coverage_map=cm, cm_vmax=-50.)
```
Which gives the following visualization:
![Figure 1](https://github.com/NVlabs/sionna/assets/43564757/146d0aa0-bf90-44ba-bcda-c1fef9c420b3)

For a steering angle of 0 degrees, you should get the following:
![Figure 2](https://github.com/NVlabs/sionna/assets/43564757/1eac53d2-2e03-4079-b005-d792da18fe69)

The steering angle is here defined with respect to the normal to the antenna array.
Note that only LoS is enabled in this example for a better visualization (``max_depth = 0``) and a vertical coverage map is defined to highlight the directivity of the radiation pattern.

Hope that helps.

**Reply by mauro-belgiovine** [reply]
Thanks for your insights and code, this is very helpful. I will look into the code you provided and will ask some further questions once I have a better understanding of the functionalities provided for directional beam steering.

**Reply by fightingsoul042** [reply]
@faycalaa Could you also share how this would work with MxN planner array to perform 3D beam steering, i.e., steering in elevation and azimuth angle both?

**Reply by mauro-belgiovine** [reply]
@fightingsoul042 you can check [my branch](https://github.com/mauro-belgiovine/sionna) to do what you asked for. I have also added an [example notebook](https://github.com/mauro-belgiovine/sionna/blob/main/examples/Sionna_Ray_Tracing_Analog_Beamsteering.ipynb) to show how the 2D vs 3D analog beamforming works. 

@faycalaa I am actually planning a pull request for this functionality, but I need to fix couple of things in the code due to recent changes in the sionna code base and fix the unit test. Do you think I can open the pull request anyway and further discuss with sionna team how to integrate it in the main repo?

**Reply by fightingsoul042** [reply]
@mauro-belgiovine Thank you for sharing this. 

@faycalaa @mauro-belgiovine I would like to ask, is it possible to use the precoding vector to get the directionality gains for each path when using paths.cir() to get the path coefficients "a"?

**Reply by faycalaa** [reply]
Hi @fightingsoul042 ,

The path coefficients  are computed for every receive and transmit antenna (`num_rx_ant` and `num_tx_ant` dimensions of [`Paths.a`](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Paths.a), respectively). You could therefore apply steering vectors to compute the directionality gains of every path.

**Reply by fightingsoul042** [reply]
@faycalaa Sorry, I didn't quite get how you meant. Can you please elaborate further.

**Reply by faycalaa** [reply]
The following code snippet illustrates how to plot the directionality gain vs the steering angle of the precoding vector.

Note that the following code only considers the LoS path and a linear array. You could do the same compute on a multiple paths (the `num_paths` dimension would then be > 1).

```py
scene = load_scene(sionna.rt.scene.munich)

num_cols = 64
scene.tx_array = PlanarArray(num_rows=1,
            num_cols=num_cols,
            vertical_spacing=0.5,
            horizontal_spacing=0.5,
            pattern="iso",
            polarization="V")

scene.rx_array = PlanarArray(num_rows=1,
                            num_cols=1,
                            vertical_spacing=0.5,
                            horizontal_spacing=0.5,
                            pattern="iso",
                            polarization="V")

# Transmitter
tx = Transmitter(name="tx",
                position=[35.11, 60.08, 10.],
                orientation=[0,0,0])
scene.add(tx)

# Receiver
rx = Receiver(name="rx",
              position=[95., 62., 10.0],
              orientation=[0,0,0])
scene.add(rx)
tx.look_at(rx)

# Rotate the tx by 45 degrees with respect to the rx
tx.orientation += np.array([0.25*np.pi, 0., 0.])

# Compute the paths
# Only LoS is enabled, and therefore only one path is computed.
paths = scene.compute_paths(max_depth=0)
a,_ = paths.cir()
# Shape of a : (batch_size = 1, num_rx = 1, num_rx_ant = 1, num_tx = 1, num_tx_ant = 64, num_paths = 1, num_time_steps = 1)

# Remove useless dims
a = tf.squeeze(a, axis=(0, 1, 2, 3, 5, 6))
# shape of a : (num_tx_ant = 64)

# Directivity gains
steeting_angles = tf.linspace(-0.5*np.pi, 0.5*np.pi, 100)
steeting_angles = tf.cast(steeting_angles, tf.float32)
dir_gains = []
for steering_angle in steeting_angles:
    # Steering vector
    precoding_vec = tf.range(num_cols, dtype=tf.float32)
    precoding_vec = tf.exp(tf.complex(0., 2.*np.pi*precoding_vec*0.5*tf.math.sin(steering_angle)))
    precoding_vec = precoding_vec / tf.linalg.norm(precoding_vec)
    # Directionality gain
    g = tf.abs(tf.reduce_sum(a*precoding_vec))
    g_db  = 10.*np.log10(g.numpy())
    dir_gains.append(g_db)

# Plot the directivity pattern
fig = plt.figure()
plt.polar(steeting_angles.numpy(), dir_gains)
fig.axes[0].set_rmin(-80)
fig.axes[0].set_theta_zero_location("N")
fig.axes[0].set_theta_direction(-1)
```

![Figure](https://github.com/NVlabs/sionna/assets/43564757/5864fe95-22fc-42f4-bb5c-4e1c8d82e3db)

**Reply by fightingsoul042** [reply]
Thank you for sharing this.

**Reply by IBRAAHIIMM** [reply]
Hi,
I was trying to implement a similar task using a uniform planar array in the TX (Ny x Nx) and a single antenna element in the receiver. The scenario is the same as in the provided code above. I generated the channel response for azimuth and elevation angles via the dot product of channel matrix and respective phase shifts from the steering vector formula for 2D arrays (sin(theta)sin(phi) for vertical axis elements and sin(theta)cos(phi) for the horizontal axis). The simulations I did for 

1.  max_depth = 0. I generated time channels, OFDM channels, (a, tau) for 6, 12, 18, and 24Ghz with 100MHz bandwidths. I checked the responses with time channels, OFDM channels, and "a" values and there is a rotation in the peak of the channel for different frequencies. However, this should not happen since I only consider the LOS path, and the receiver location is fixed.

2. max_depth = 5. The same issue happens in this case too. 

Do you have any insight into why does this happen? Ideally, it should not happen since the receiver is very close to the TX, and even if there are reflections, the LOS path should be dominant. The peak should occur at the similar elevation and azimuth angle for different frequencies.

**Reply by faycalaa** [reply]
Hi,
Could you please share a minimal code snippet that reproduces the problem?

**Reply by IBRAAHIIMM** [reply]
Hi, 
Sorry for the late reply. I could not get back to you. Thanks for the reply! Here is the code snippet.

```py
#Angular domain resolution
N_azi = 20
N_ele = 20
#Antenna config
Nx = 8  # Number of elements along the horizontal axis
Ny = 8  # Number of elements along the vertical axis
dx = 0.5  # Spacing between elements along the x-axis
dy = 0.5  # Spacing between elements along the y-axis
wavelength = 1 

#define the simulation range
azimuth_angles = np.linspace(-np.pi/2, np.pi/2, N_azi)  # Azimuth angles from -90 to 90 degrees
elevation_angles = np.linspace(-np.pi/2, np.pi/2, N_ele)  # Elevation angles from -90 to 90 degrees
beamforming_matrix = np.zeros((len(azimuth_angles), len(elevation_angles), Ny, Nx), dtype=complex)

#Calculate the positions of the elements in the UPA
x_positions = np.arange(Nx) * dx
y_positions = np.arange(Ny) * dy

for az_idx, azimuth in enumerate(azimuth_angles):
    for el_idx, elevation in enumerate(elevation_angles):
        # Calculate the wave vector
        k = 2 * np.pi / wavelength * np.array([np.sin(elevation) * np.cos(azimuth),np.sin(elevation) * np.sin(azimuth),np.cos(elevation)])
        # Loop through each element position
        for ny, y_pos in enumerate(y_positions):
            for nx, x_pos in enumerate(x_positions):
                # Calculate the phase shift for the element
                phase_shift = np.exp(-1j * np.dot(k, [x_pos, y_pos, 0]))
                # Store the beamforming vector in the 4D matrix
                beamforming_matrix[az_idx, el_idx, ny, nx] = phase_shift
                
#Reshape the beamforming vectors into column vectors based on the antenna element representation in the Sionna
beamforming_matrix = np.transpose(beamforming_matrix,(0,1,3,2)).reshape(beamforming_matrix.shape[0],beamforming_matrix.shape[1],Nx*Ny)

"""Get the time channel or the frequency channel for """


scene = load_scene(sionna.rt.scene.munich)

Ny = 8 #vertical elements
Nx = 8 #horizontal elements
fc = 24e9
bandwidth =100e6
max_depth = 0

scene.tx_array = PlanarArray(num_rows=Ny,
            num_cols=Nx,
            vertical_spacing=0.5,
            horizontal_spacing=0.5,
            pattern="iso",
            polarization="V")

scene.rx_array = PlanarArray(num_rows=1,
                            num_cols=1,
                            vertical_spacing=0.5,
                            horizontal_spacing=0.5,
                            pattern="iso",
                            polarization="V")


 #Create transmitter
tx = Transmitter(name="tx",
              position=[8.5,21,27],
              orientation=[0,0,0])
scene.add(tx)

# Create a receiver
rx = Receiver(name="rx",
          position=[45,90,1.5],
          orientation=[0,0,0])
scene.add(rx)


tx.look_at(rx)

#Rotate the tx by 45 degrees with respect to the rx
tx.orientation += np.array([0.45*np.pi,0.0, 0.])


#set the frequency in the scene
scene.frequency = fc

paths = scene.compute_paths(max_depth=max_depth, num_samples=5e5, diffraction=False, scattering=False)#, scat_keep_prob=scat_prob)


a,tau = paths.cir()


l_min,l_max = sionna.channel.time_lag_discrete_time_channel(bandwidth, maximum_delay_spread=3e-06)
print(f"fc {fc} Number of channel taps {l_max-l_min}")
h_time = sionna.channel.cir_to_time_channel(bandwidth, a, tau, l_min, l_max, normalize=False)


"""Calculate the channel gain"""
#initialize 2D matrix where columns are azimuth angle responses and rows are elevation angle responses
channel_gain  = np.zeros((N_ele,N_azi))
for i_azi in range(N_azi):
    for j_ele in range(N_ele):
        beamformer = beamforming_matrix[i_azi,j_ele].reshape([Nx*Ny,1])
        beamformed_channel_taps = np.einsum("ij,ic->jc",beamformer,h_time)
        channel_gain[j_ele,i_azi] = np.sum(np.abs(beamformed_channel_taps)**2)
```
I am plotting the heatmap of channel_gain. There seems to be a rotation in the peak for different frequencies.

**Reply by IBRAAHIIMM** [reply]
I do not know why it does not preview the code.

**Reply by faycalaa** [reply]
Hi @IBRAAHIIMM ,

You should set the scene frequency *before* configuring the antenna arrays to ensure that the correct spacing is used by the antenna arrays. In your current code, the scene frequency is set after setting the antenna arrays, and therefore the antenna spacing of the arrays is set using the wavelength corresponding to the default carrier frequency (3.5 GHz).

**Reply by IBRAAHIIMM** [reply]
Hi @faycalaa, 
Thanks for your input! It is fixed now.

**Reply by jhoydis** [reply]
Hi @IBRAAHIIMM,

Since the latest release (v0.18), this problem should be automatically fixed because the spacing of the elements of a planar array now scale automatically with the scene frequency.
